### ISSUE TITLE
bindAndHandleWith unbinds if a connection fails (which can be caused by client)

### ISSUE THREAD

**Issue by Joe-Edwards** [akka/akka#17992]
If you create a simple server using `Http().bindAndHandleWith`, any failure in the connection stream will cause the connection source to be cancelled, which subsequently causes the server to unbind.

The problem appears to be the mapAsyncUnordered at Http.scala#L97. The future here comes from the 'identityFinishReporter', which fails when a connection stream fails.

This comes before the handler is used, so user code cannot produce this behaviour, but the TCP stream can! In particular, when a connection is reset by a client unexpectedly a `akka.stream.StreamTcpException: The connection closed with error An existing connection was forcibly closed by the remote host` is sometimes* produced.

In other words, a badly behaved client can take down the server by forcibly closing a single connection!

*I couldn't produce this particularly reliably, I don't know exactly when a connection is deemed to have completed gracefully as opposed to forcibly.

**Comment by jrudolph** [comment]
@Joe-Edwards thanks for reporting. If we can reproduce it, I agree it would be quite bad.

**Comment by Joe-Edwards** [comment]
The test I was using to reproduce it is https://gist.github.com/Joe-Edwards/b35373f49b16c7656640

It runs 100 times because I don't have a reliable way of producing the stream exception. (Perhaps I need to do something more subtle like failing a request halfway through it's body or something)

**Comment by jrudolph** [comment]
From looking at the code I think that your summary is correct. We can probably just ignore errors coming out of the `incomingConnection.flow`.

**Comment by jrudolph** [comment]
@Joe-Edwards You are on Windows, right? The `An existing connection was forcibly closed by the remote host` seems to correspond to the well-known `Connection reset by peer` error as occurred on other platforms.

I haven't managed to reproduce the exact issue so far but AFAIU it would happen if the Akka.io socket gets `ErrorClosed` for some reason like RSTing the connection. However, even then, there's a race provoked by the TCP stream implementation between reporting the error on the network input side against cancelling the network output side. If cancelling is faster nothing bad will happen. If reporting the error is faster we close the binding socket. I could never get the error to be faster than the cancellation (but I also haven't tried to understand the exact call-path by which the network output side is cancelled).

So, for now I will submit a fix without a reproduction that seems hard to find.

/cc @sirthias

**Comment by jrudolph** [comment]
Fix in #17996.

**Comment by ktoso** [comment]
Thanks for reporting @Joe-Edwards and the fix @jrudolph !

**Comment by Joe-Edwards** [comment]
Yeah, I'm developing on Windows, though I first noticed this when running my app on Linux. With even a fairly light load from a badly behaved client, it would mysteriously stop responding to requests almost immediately!

I had much the same thoughts about the cancellation racing back up. None of the things I tried made it more reliable - it just happened occasionally.

Fix and the new ticket look good. Thanks.
